crates/typst-library/src/foundations/dict.rs

```diff
@@ -254,6 +256,101 @@ impl Dict {
             .map(|(k, v)| Value::Array(array![k.clone(), v.clone()]))
             .collect()
     }
+
+    /// Produces a new dictionary or array by transforming each key-value pair with the given function.
+    ///
+    /// If the mapper function returns a pair (array of length 2), the result will be a new dictionary.
+    /// Otherwise, the result will be an array containing all mapped values.
```

Comment:
Perhaps this should be a different function: you might want to return an array even when the function returns or might return an array of length 2.